File: SeriesGuide/src/main/java/com/battlelancer/seriesguide/service/NotificationActionReceiver.java
package com.battlelancer.seriesguide.service;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.support.v4.app.NotificationManagerCompat;
import com.battlelancer.seriesguide.SgApp;
import com.battlelancer.seriesguide.enums.EpisodeFlags;
import com.battlelancer.seriesguide.provider.SeriesGuideContract;
import com.battlelancer.seriesguide.util.EpisodeTools;

/**
 * Listens to notification actions, currently only setting an episode watched.
 */
public class NotificationActionReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {
        if (NotificationService.ACTION_CLEARED.equals(intent.getAction())) {
            NotificationService.handleDeleteIntent(context, intent);
            return;
        }

        int episodeTvdbvId = intent.getIntExtra(NotificationService.EXTRA_EPISODE_TVDBID, -1);
        if (episodeTvdbvId <= 0) {
            return; // not notification set watched action
        }

        // query for episode details
        Cursor query = context.getContentResolver()
                .query(SeriesGuideContract.Episodes.buildEpisodeWithShowUri(episodeTvdbvId),
                        new String[] {
                                SeriesGuideContract.Shows.REF_SHOW_ID,
                                SeriesGuideContract.Episodes.SEASON,
                                SeriesGuideContract.Episodes.NUMBER }, null, null, null);
        if (query == null) {
            return;
        }
        if (!query.moveToFirst()) {
            query.close();
            return;
        }
        int showTvdbId = query.getInt(0);
        int season = query.getInt(1);
        int episode = query.getInt(2);
        query.close();

        // mark episode watched
        EpisodeTools.episodeWatched(context, showTvdbId, episodeTvdbvId, season, episode,
                EpisodeFlags.WATCHED);

        // dismiss the notification
        NotificationManagerCompat manager = NotificationManagerCompat.from(context);
        manager.cancel(SgApp.NOTIFICATION_EPISODE_ID);
        // replicate delete intent
        NotificationService.handleDeleteIntent(context, intent);
    }
}

File: app/src/main/java/com/battlelancer/seriesguide/notifications/NotificationActionReceiver.kt
// SPDX-License-Identifier: Apache-2.0
// Copyright 2016-2018, 2021-2024 Uwe Trottmann

package com.battlelancer.seriesguide.notifications

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import com.battlelancer.seriesguide.shows.episodes.EpisodeFlags
import com.battlelancer.seriesguide.shows.episodes.EpisodeTools

/**
 * Listens to notification actions, currently only setting an episode watched.
 */
class NotificationActionReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        if (NotificationService.ACTION_CLEARED == intent.action) {
            NotificationService.handleDeleteIntent(context, intent)
            return
        }
        val episodeId = intent.getLongExtra(EXTRA_LONG_EPISODE_ID, 0)
        if (episodeId <= 0) {
            return  // not notification set watched action
        }

        // mark episode watched
        EpisodeTools.episodeWatched(context, episodeId, EpisodeFlags.WATCHED)

        // dismiss the notification
        NotificationService.deleteNotification(context, intent)
    }

    companion object {

        private const val EXTRA_LONG_EPISODE_ID = "episode_id"

        fun intent(episodeId: Long, context: Context): Intent {
            return Intent(context, NotificationActionReceiver::class.java)
                .putExtra(EXTRA_LONG_EPISODE_ID, episodeId)
        }
    }
}